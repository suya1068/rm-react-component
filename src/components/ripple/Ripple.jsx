import "./Ripple.scss";
import classnames from "classnames";
import React, { Component } from "react";
import { autobind } from "core-decorators";

@autobind
class Ripple extends Component {
    state = {
        isRunning: false
    };

    onClick(event) {
        this.showRipple();
    }

    onAnimationEnd(event) {
        this.hideRipple();
    }

    showRipple() {
        this.setState({ isRunning: true });
    }

    hideRipple() {
        this.setState({ isRunning: false });
    }

    render() {
        const { isRunning } = this.state;

        return (
            <div
                className={classnames("u-ripple", { "is-active": isRunning })}
                onClick={this.onClick}
                onAnimationEnd={this.onAnimationEnd}
            ><div className="u-ripple-circle" /></div>
        );
    }
}

export default Ripple;
