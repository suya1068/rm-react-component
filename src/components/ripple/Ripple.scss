.u-ripple {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    overflow: hidden;
    background: transparent;
}

.u-ripple-circle {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    opacity: 0;
    width: 0;
    height: 0;
    border-radius: 50%;
    background: rgba(255, 255, 255, 0.25);

    .u-ripple.is-active & {
        animation: ripple 0.4s ease-in;
    }
}

@keyframes ripple {
    0% {
        opacity: 0;
    }
    25% {
        opacity: 1;
    }
    100% {
        width: 200%;
        padding-bottom: 200%;
        opacity: 0;
    }
}
