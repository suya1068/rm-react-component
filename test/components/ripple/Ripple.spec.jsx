import React from "react";
import Ripple from "src/components/ripple/Ripple";

describe("Ripple Component", () => {
    it("랜더링한다.", () => {
        const component = shallow(<Ripple />);
        expect(component.state().isRunning).toBe(false);
        expect(component).toMatchSnapshot();
    });

    describe("클릭하면, 잔물결 모양을 보여준 후 300ms 후 사라진다.", () => {
        it("잔물결 모양을 보여준다.", () => {
            const component = shallow(<Ripple />);
            component.instance().showRipple();
            expect(component.state().isRunning).toBe(true);

            component.update();
            expect(component).toMatchSnapshot();
        });

        it("잔물결 모양을 숨긴다.", () => {
            const component = shallow(<Ripple />);
            component.instance().hideRipple();
            expect(component.state().isRunning).toBe(false);

            component.update();
            expect(component).toMatchSnapshot();
        });
    });
});
