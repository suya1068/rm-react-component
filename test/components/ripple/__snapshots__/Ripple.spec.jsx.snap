// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple Component 랜더링한다. 1`] = `
<div
  className="u-ripple"
  onAnimationEnd={[Function]}
  onClick={[Function]}
>
  <div
    className="u-ripple-circle"
  />
</div>
`;

exports[`Ripple Component 클릭하면, 잔물결 모양을 보여준 후 300ms 후 사라진다. 잔물결 모양을 보여준다. 1`] = `
<div
  className="u-ripple is-active"
  onAnimationEnd={[Function]}
  onClick={[Function]}
>
  <div
    className="u-ripple-circle"
  />
</div>
`;

exports[`Ripple Component 클릭하면, 잔물결 모양을 보여준 후 300ms 후 사라진다. 잔물결 모양을 숨긴다. 1`] = `
<div
  className="u-ripple"
  onAnimationEnd={[Function]}
  onClick={[Function]}
>
  <div
    className="u-ripple-circle"
  />
</div>
`;
